### ISSUE TITLE
Transfer-Encoding: chunked HTTP response - missing data

### ISSUE THREAD

**Issue by aleksvujic** [traefik/traefik#9775]
# What did you do?

## Architecture diagram
We are using Apache Thrift RPC server written in C#. It is exposed behind Traefik and Kestrel over HTTP protocol. Architecture diagram looks like this:

![Traefik and Thrift](https://user-images.githubusercontent.com/17112542/224711860-08ac0a22-5aed-422f-8961-819b0f54e4c3.png)

We noticed that sometimes not all the data returned by the server comes through to the Python client.

## Investigation and related issues
The problem occurs only when responses are chunked. If the `Content-Length` HTTP header is set, everything works as expected.
1. Chunked response
    1. Responses are chunked because of the design of ASP.NET Core servers (see issue [2417](https://github.com/aspnet/KestrelHttpServer/issues/2417)).
    2. `UseResponseBuffering()` is deprecated and will be maybe implemented in .NET 8 (see issue [27470](https://github.com/dotnet/aspnetcore/issues/27470)).
    3. Workaround, if Traefik does not fix the issue, would be to set the `Content-Length` HTTP header (see issue [17413](https://github.com/dotnet/aspnetcore/issues/17413)).
2. No option in Traefik seems to fix this ([Middlewares - Buffering](https://doc.traefik.io/traefik/middlewares/http/buffering/)).
3. Only similar issue: [Traefik ingress route on k8s, modifies chunked response from service](https://community.traefik.io/t/traefik-ingress-route-on-k8s-modifies-chunked-response-from-service/9013)

## Attachment
[`traefik-report.zip`](https://github.com/traefik/traefik/files/10958588/traefik-report.zip) - contains all the necessary files to reproduce the issue and a Wireshark recording

1. `traefik` folder - Traefik configuration
2. `thrift_server_netstd` folder - implementation and deployment files for C# server
3. `thrift_client_py` folder - implementation and deployment files for Python client
4. `traefik-bug.pcapng` - Wireshark capture of the network traffic when the issue occurred

## Steps to reproduce

1. Deployment machine `depl1` info:
```
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

Docker Version 20.10.21, build baeda1f
```
2. Unzip [`traefik-report.zip`](https://github.com/traefik/traefik/files/10958588/traefik-report.zip) (attachment of this issue) on the deployment machine `depl1` and test machine `test1`.
3. Go to `traefik-report/traefik` folder and run the following command on `depl1`:
```sh
docker stack deploy -c docker-compose-Traefik.yml Traefik
```
4. Go to `traefik-report/thrift_server_netstd` folder and deploy Thrift server on `depl1`:
```sh
docker build -t thrift-server -f Dockerfile .
docker stack deploy -c docker-compose.yml thrift-service
```
5. Add a new record to `hosts` file which resolves `thrift-service.com` on `test1` to the IP where `thrift-service` is deployed (`depl1`).
6. Go to `traefik-report/thrift_client_py` folder and run client in Python environment with [Thrift library](https://pypi.org/project/thrift/) on `test1` (_it does not always fail and maybe a couple of retries are necessary_):
```
python3 thrift_client.py
```

# What did you see instead?

Sometimes not all the data that the C# server sends back comes through to the client. This is the error that we receive in the Python client:

```
Traceback (most recent call last):
  File "C:\Users\dejank\Desktop\dev\traefik-report\thrift_client_py\thrift_client.py", line 22, in <module>
    client.read_data(strData)
  File "C:\Users\dejank\Desktop\dev\traefik-report\thrift_client_py\gen_py\helloWorld\helloWorldInterface.py", line 45, in read_data
    return self.recv_read_data()
  File "C:\Users\dejank\Desktop\dev\traefik-report\thrift_client_py\gen_py\helloWorld\helloWorldInterface.py", line 64, in recv_read_data
    result.read(iprot)
  File "C:\Users\dejank\Desktop\dev\traefik-report\thrift_client_py\gen_py\helloWorld\helloWorldInterface.py", line 209, in read
    self.success.read(iprot)
  File "C:\Users\dejank\Desktop\dev\traefik-report\thrift_client_py\gen_py\helloWorld\ttypes.py", line 58, in read
    (fname, ftype, fid) = iprot.readFieldBegin()
  File "C:\isys_venv\lib\site-packages\thrift\protocol\TBinaryProtocol.py", line 163, in readFieldBegin
    type = self.readByte()
  File "C:\isys_venv\lib\site-packages\thrift\protocol\TBinaryProtocol.py", line 207, in readByte
    buff = self.trans.readAll(1)
  File "C:\isys_venv\lib\site-packages\thrift\transport\TTransport.py", line 68, in readAll
    raise EOFError()
EOFError

Process finished with exit code 1
```

# What version of Traefik are you using?

```
Version:      2.9.8
Codename:     banon
Go version:   go1.19.6
Built:        2023-02-15T15:23:25Z
OS/Arch:      linux/amd64
```

# What is your environment & configuration?

All the relevant Traefik configuration is also included in [`traefik-report.zip`](https://github.com/traefik/traefik/files/10958588/traefik-report.zip) attachment in `traefik` folder.

`docker-compose-Traefik.yml`:
```yml
version: "3.7"

services:
  traefik:
    image: "traefik:v2.9.8"
    networks:
      - traefik-net
    ports:
      - target: 9090
        published: 9090
        protocol: tcp
        mode: host
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock:ro"
      - "./TraefikConfig/:/etc/traefik/"
    deploy:
      replicas: 1
      labels:
        - "traefik.enable=true"
        - "traefik.docker.network=traefik-net"

networks:
  traefik-net:
    external: true
    name: traefik-net
```

`traefik.yml`:
```yml
serversTransport:
  insecureSkipVerify: true

providers:
  file:
    directory: "/etc/traefik"
    watch: true
  docker:
    swarmMode: true
    exposedByDefault: false

entrypoints:
  thrift-service:
    address: ":9090"

```

**Comment by meganoob1337** [comment]
I think i have a similar Problem. Although just plain Docker , and a Symfony application behind the Traefik.
Sometimes the Json response is just truncated, but not always and not always at the same spot.
We are using version 2.5.7
I found this for v1.5 https://github.com/traefik/traefik/pull/2644 but i dont know if it relates to this. 

we also dont use chunks (im not that knowledgeable in traefik and http requests) so i guess i should open a new Bug report?
the Data our Backend is sending is Complete though.

Could a missing Content-Length header have anything to do with it?
Other Requests are running fine though - but this is the biggest with around 225kb (without gzip)

**Comment by aleksvujic** [comment]
Any update on this issue? Were you able to reproduce it?

**Comment by meganoob1337** [comment]
At least for me (not the OP) it was a problem in our backend application. Our nginx had a wrong permission on the folder he would cache big responses. So as soon as a response was bigger than the allowed memory buffer he would try to write it - fail - and we would get a non-compete response . Maybe this helps anyone

**Comment by aleksvujic** [comment]
Thank you so much for investing your time into researching this issue. Please keep us informed on your progress. If you need any additional details, let us know.

**Comment by aleksvujic** [comment]
Any update on this issue?

**Comment by cdemi** [comment]
I believe this issue is related to #7930. I think it's still present in 2.10.5.

In summary, whenever a buffering middleware is present, `Transfer-Encoding: chunked` responses will be built on traefik and flushed at the end

**Comment by thomafred** [comment]
We are experiencing similar issues.

We have a uvicorn-based web-api which produces a response using chunked transfer encoding. Port-forwarding into the server works without issues, but going through our traefik ingress controller leaves us with a truncated response.

**Comment by lemniskett** [comment]
We also experiencing a similar issue.

We deploy Odoo (gunicorn-based) behind traefik as LB, the database backup feature in Odoo produces a chunked response and when going through traefik the response is truncated.

**Comment by nerdalertdk** [comment]
Hi

We have some older CPE devices that also don't like how traefik is handling "Transfer-Encoding"
directly to our apache server it works fine, since apache returns "Content-Length" and don't chunk the xml data.

an option just to for  Content-Length over Transfer-Encoding would fix my problem :)

**Comment by MoonKON** [comment]
Hi! Do we have any useful solution for this issue? I think traefik also have some issues on the chunk encoding in aim remote repo. We have an aim development in k8s cluster. When we try to access the remote repo in the aim through traefik, the chunk encoding issues happens. Messages cannot be chunked correctly.

**Comment by tonybart1337** [comment]
I have the following use case:
Client sends a HTTP/1.1 request to the server and downloads a file which is dynamically generated and returned in chunks, but sometimes the connection might abort midway and instead of aborting the connection traefik sends zero length chunk finalizing the request and the client ends up receiving a truncated response with no way of knowing whether it received the full file or not.

I think this is similar to this issue: https://github.com/traefik/traefik/issues/10858, I might be wrong.
I took a little dive into the source code and it looks like the issue lies down in the recovery middleware.
If I simply comment out this line https://github.com/traefik/traefik/blob/v3.1.4/pkg/middlewares/recovery/recovery.go#L30 then traefik successfully aborts the request and the client receives the error

@rtribotte maybe that would help you narrow it down:
```bash
traefik-1  | 2024-09-24T22:07:35Z DBG usr/local/go/src/log/log.go:245 > 2024/09/24 22:07:35 reverseproxy.go:677: httputil: ReverseProxy read error during body copy: unexpected EOF
traefik-1  | 172.16.16.1 - - [24/Sep/2024:22:07:35 +0000] "GET /download HTTP/1.1" 200 2048 "-" "-" 1 "to-test@file" "http://10.12.10.8:58239" 42m
```

**Comment by tonybart1337** [comment]
[PR](https://github.com/traefik/traefik/pull/11129)

**Comment by rtribotte** [comment]
Fixed by #11129
